### ISSUE TITLE
Required reviewers users/groups should not be automatically added to reviewers list

### ISSUE THREAD

**Issue by mnhorak** [monitorjbl/pr-harmony#13]
When I specify required reviewers, they are automatically added to the reviewers list of new pull requests as if they were also default reviewers. It would be more useful for my company's workflow if required reviewers were not also default reviewers.

From the settings page:
Default reviewers: Groups that are **added** to approve every pull request (comma-separated)
Required reviewers: Groups that are **required** to approve every pull request (comma-separated)

Thanks for writing this add-on. It would be that much more awesome if this issue could be addressed!

**Comment by monitorjbl** [comment]
Hm, I'm not sure about that, and here's why. The required reviewers are currently always be added to the PR so they can approve it. It would be kind of a pain for required reviewers to have to manually add themselves to PRs. Plus they wouldn't get notified about the PR, so they'd have to check periodically to see if there are new requests to review. 

It wouldn't be too hard to pull them out of the create screen so the user opening the PR wasn't burdened by having to see reviewers he couldn't really change. Would that work for you? The consequence is that there will be a bit of a delay from when the PR is opened to when the required users are added. It's an async event, so the user would have to refresh the page to see them in some cases.

**Comment by mnhorak** [comment]
After reading your explanation, I realized that this add-on functions differently than I thought. Required reviewers were added to the PR even if I deleted them from the form when creating the PR. They appeared when I refreshed the page.

I am very interested in the "# of reviews" feature only and specifying the subset of users/groups whose approval adds to this total. I am not interested in having this set of users added to and notified of _every_ pull request. Maybe these are "required approval reviewers".

The friction for my team is the amount of emails the required reviewers would receive for PR's that don't concern them.

I suppose you can close this issue as invalid, since I misunderstood the feature. If you are interested in adding the functionality I described above, I could write a new issue.

Thanks for the quick response and support!

**Comment by monitorjbl** [comment]
Ah, I see. I suspect you have most of your code one repo, which isn't necessarily a bad thing but it does complicate reviews a bit. Where I work, we keep separate pieces of our application in separate repos and combine them with Google's [repo tool.](https://code.google.com/p/git-repo/)  This makes choosing reviewers for pull request pretty straightforward because we can just delegate people to each repo. 

Your use case is more complex because there's no way to inform the user in the create UI what the pool of reviewers should be. The best you could do is give them an error message after they click Create informing them of what users they could pick from. 

I think if this were to be implemented it would be as a checkbox in the config screen that specified if required reviewers are added automatically. That way the current process could exist in parallel; if the repo specified that required of us were not added automatically the user would have to provide them. Not super user friendly but I think it satisfies your requirement.

**Comment by mnhorak** [comment]
The checkbox on the config screen to disable auto-adding required reviewer users and groups would satisfy our needs. That would be awesome.

**Comment by rogerbramon** [comment]
I was expecting the same like @mnhorak. We would like to force a group of users to be reviewed for one reviewer of another group but we don't want to notify all of them every PR. 

The checkbox in the config screen would be enough for us. 

Another option would be to notify when creating a pull request that X reviewers from a given list of users must be added.

Thanks for your work!

**Comment by monitorjbl** [comment]
I've been messing around with this process and the more I have the more I think your original suggestion makes more sense for all cases @mnhorak. If there are not at least N number of required reviewers added to the PR, there should be a warning. If anyone wants to make sure that no intervention is needed on the part of the user, they can just add the required reviewers to the default reviewers as well.

**Comment by monitorjbl** [comment]
Here are a couple screenshots of what I've got implemented now. Let me know if you have any suggestions for the workflow. I'll plan on releasing to the marketplace it at the end of the week unless I hear otherwise.

So, for this configuration:

![image](https://cloud.githubusercontent.com/assets/1149136/9297513/b7ffb1d4-446e-11e5-9fc5-695a83cc0cc1.png)

A user will see the following when they go to open a PR:

![image](https://cloud.githubusercontent.com/assets/1149136/9297492/5c795748-446e-11e5-9699-5c3bae0c83eb.png)

If two of the required reviewers are added as default reviewers to the config, the popup won't show at all and the user can immediately create the PR. Otherwise, the user will have to add required reviewers at his own discretion.

**Comment by rogerbramon** [comment]
Hey! That sounds good. 
Is there any way to have a beta version to try it and give you feedback?

My first thought is that this new approach fits very well for reviewer groups but for reviewer users the previous one fits better. If these users are required, make sense to include them from the beginning.

**Comment by monitorjbl** [comment]
Sure, I can put a beta release up on GitHub. I don't think this process changes the reviewer groups workflow too much. In the old version, you could still make the default group and then the required reviewers would be there by default.  

The only restriction this imposes is that you can't open a PR that doesn't have enough required reviewers on it to begin with. The old method just made sure that all required reviewers always were added to the PR after it was opened, regardless of what the user chose. 

If that workflow is desired, you just have to set the requiredReviews to the number of requiredReviewers you have. And if you don't want to make your users select those reviewers, then you can simply set the defaultReviewers to be the same as the requiredReviewers.

**Comment by mnhorak** [comment]
Taylor, thanks for your work on the issue! I'm excited to try it out when you release it.

**Comment by monitorjbl** [comment]
Sorry for the delay @rogerbramon and @mnhorak , I meant to put this up last night but got home pretty late. The beta version is now available here: https://github.com/monitorjbl/pr-harmony/releases/tag/v1.7-RC1. Let me know what you think!

**Comment by rogerbramon** [comment]
Thanks Taylor. I tried it and I got a couple of problems. I'm on Stash 3.7.0.
- Default users and groups are not added to the list.
- I can create a pull request without adding the required users. The warning isn't popping up.

On the other hand, I don't know if it's a bug, but when I create a PR and I'm part of the required reviewers group, I can merge it with no approvals if only an approval is required.

**Comment by monitorjbl** [comment]
For the first two, that's odd...can you give me a screenshot of your config? Want to make sure there's nothing unexpected in there. I'll take a look tonight and see if I can replicate it locally.

For the last one, that's actually a feature as long as you have Required Reviews set to 1. If you are the only required reviewer and you open a PR, it would be kind of a pain if you couldn't merge it :)

**Comment by rogerbramon** [comment]
![pr-harmony](https://cloud.githubusercontent.com/assets/330441/9388428/9a398d92-4765-11e5-9528-a5111700082d.png)

**Comment by monitorjbl** [comment]
Is that user the one you're opening the PR as?

**Comment by rogerbramon** [comment]
No, it's another one.

**Comment by mnhorak** [comment]
I have one issue that I can reproduce.
- Add a required group (group1) and set number of required reviewers to 1
- Create pull request with 2 users, 1 in group1 (required-user) and 1 not in group1 (regular-user)
- Have regular-user approve the PR
- Observe that required-user can now merge the PR even though no required reviewers approved

I do see the warning about required reviewers when creating the PR. And I cannot create the PR without  a required approver.

**Comment by monitorjbl** [comment]
@rogerbramon I was able to replicate this. It looks like this will break compatibility with older versions of Stash. The AUI elements being used are only available from 3.9.0 forward :(

**Comment by monitorjbl** [comment]
@mnhorak Did you open the PR as required-user? If so, he will be able to merge it without approval, as he is the only required reviewer and it would be pretty bad if he couldn't ever merge his PRs. If you're opening the PR as a third user, then I can't replicate this in Stash 3.9.0+. What version are you using?

I do see an issue with the UI blocking PRs in that instance though, I'll get that fixed.

**Comment by mnhorak** [comment]
In my case, group1 has multiple users beyond required-user. The user that created the PR is in group1, but is not required-user.

Agreed there is a special case when there is only one required reviewer total since they can't approve their own PR.

**Comment by monitorjbl** [comment]
Ah, then I think you hit the same case. Let's pretend for the moment that you had 2 required reviewers (userA, and userB) and a required review count of 1. If userA opens a PR, it will be able to be merged immediately as userA's submission effectively counts as an approval because as a required reviewer he is "trusted". The group-based case you're describing is more or less identical, so I think it's working as intended.

**Comment by monitorjbl** [comment]
I should add that if you feel like that logic should operate differently, it certainly could change but I'd like it to be a separate issue.

**Comment by mnhorak** [comment]
OK I think I had the Stash built-in check for 1 approver enabled. That would make sense why one approval from any user would activate the merge.

On my team, we have a "no one is above the law" mentality, so having some users able to merge without another team member's approval would be undesirable.

Agreed that can be a separate issue. Thanks for your work on this issue and your awesome responsiveness!

**Comment by rogerbramon** [comment]
My team has the same policy. Right now we're using the built-in check for approvals to force this policy. I agree to treat it as a separate issue. Thanks a lot!

**Comment by rogerbramon** [comment]
@monitorjbl let me know if you need I do some testing on Stash 3.7 before you release the new version. Thanks!

**Comment by monitorjbl** [comment]
@rogerbramon I don't think this can be made compatible with Stash 3.7 without writing custom styles for the PR create screen. The AUI elements I'm using are only available on Stash 3.9.0+.  Do you have any plans to upgrade?

**Comment by rogerbramon** [comment]
@monitorjbl OK, so 1.7 will be compatible with Stash 3.9.0+. I'll ask for the plans to upgrade.

**Comment by monitorjbl** [comment]
Yeah, specifically it's this component that isn't available in 3.7: https://docs.atlassian.com/aui/latest/docs/flag.html.

**Comment by monitorjbl** [comment]
@mnhorak @rogerbramon I opened #16 to track the logic change you guys requested. Have a look i you get a chance and I'll get to work on it. Thanks so much for beta testing the plugin, I really appreciate your feedback!

**Comment by rogerbramon** [comment]
Thanks, @monitorjbl I'll comment on that. 
Regarding this issue, I don't understand why aui-flag doesn't work on Stash 3.7. Theoretically Stash 3.7 uses AUI 5.7.19 ([link](https://bitbucket.org/atlassian/aui-adg/wiki/versions/product-version-matrix)) and aui-flag is experimental since 5.7 and it is included in the docs of 5.7.19 ([link](https://docs.atlassian.com/aui/5.7.19/docs/flag.html)).

The error I get when creating a PR is this:
![aui-flag-error](https://cloud.githubusercontent.com/assets/330441/9516710/c3432fd2-4caa-11e5-8593-cc5c888ff0ff.png)

I'm not familiar with Stash plugins and I would appreciate any idea on that because we really want this features but we can't update Stash right now.

Thanks!!!

**Comment by monitorjbl** [comment]
I'm honestly not sure. The Atlassian SDK let's you run any version of Stash you want, and I ran 3.7, 3.8, and 3.9 to test this out. The flag element only works past 3.9. 

It's possible there's something else going on here, because I saw the same compatibility matrix as you. Experimentally, that matrix appears to be wrong.  I asked a question on the Atlassian Q&A site, hopefully they'll clear it up for us: https://answers.atlassian.com/questions/24662165/why-does-the-aui-flag-element-not-work-in-stash-3.7

**Comment by monitorjbl** [comment]
Okay, so it turns out that I have to explicitly enable it as `<stash-resource/>` in older versions. I've done so in the [latest build](https://github.com/monitorjbl/pr-harmony/releases/tag/v1.7-RC2), can you give it a shot @rogerbramon ?

**Comment by rogerbramon** [comment]
Great! It works now! :D
Just to check if it's the expected workflow: If a required reviewers group if defined but no default reviewers, PR can be created if you don't add any reviewer.

Thanks!!!!

**Comment by mnhorak** [comment]
Uh oh. I just tried RC2 and now do not see the warning for required
reviewers when creating a PR. And can create a PR with 0 reviewers. Stash
version 3.11.2.

**Comment by monitorjbl** [comment]
Yikes, that's not good. I hope there's some magic combination of tags that initializes this thing for both versions. Really don't want to have to release two versions...

**Comment by monitorjbl** [comment]
@mnhorak I can't reproduce this using the SDK running Stash 3.11.2 in Chrome, Firefox or Safari. Are you using the same config you were using before? Here's what I'm using:

![image](https://cloud.githubusercontent.com/assets/1149136/9568994/f0a5b69e-4f29-11e5-8e28-8c51ff09f55f.png)

**Comment by mnhorak** [comment]
I had just a single reviewer group defined. I installed this on my Stash
instance where I tested RC1. When I uninstalled and installed RC1 again, I
again saw the reviewer warning.

**Comment by monitorjbl** [comment]
If the test conditions are the exact same both times, I'm wondering if the uninstall process didn't complete properly. Are there any log entries on the server side? If you know how to use your browser's dev console ([here's how to use Chrome's](https://developer.chrome.com/devtools/docs/console)), do you see any errors in that?

If there's no indication of a problem in either of those, would you be willing to have a Skype call or something to show me some details about your environment? Hate to keep dragging this out over Gihub updates.

**Comment by mnhorak** [comment]
I'll give RC2 another try tomorrow and see if there are any log messages or
browser errors.

**Comment by mnhorak** [comment]
@monitorjbl Sorry for the confusion! I retested this morning and realize this was my fault. I had the number of reviewers set to 1 and the user creating the PR was in the required reviewer group. Based on the current behavior as discussed previously, they could create the PR without the warning message appearing. This could be configured differently due to #16.
